File: KeySAVCoreJS/scripts/source/BattleVideoReader.cs
using System;
using System.IO;
using KeySAVCore.Structures;
using JS;

namespace KeySAVCore
{
    public class BattleVideoReader
    {
        private const ushort offset = 0x4E18;
        private const ushort keyoff = 0x100;

        private Uint8Array video;
        private Uint8Array key;

        internal BattleVideoReader(Uint8Array file, Uint8Array key_)
        {
            video = file;
            key = key_;
        }
        
        public PKX? getPkx(byte slot, bool opponent)
        {
            Uint8Array ekx;
            Uint8Array pkx;
            byte opponent_ = opponent ? (byte)1 : (byte)0;
            ekx = Utility.xor(video, offset + 260 * slot + opponent_ * 0x620, key, keyoff + 260 * slot + opponent_ * 0x700, 260);
            pkx = PKX.decrypt(ekx);
            if (pkx.Empty())
            {
                return null;
            }
            return new PKX(PKX.verifyCHK(pkx) ? pkx : ekx, -1, slot, false);
        }

        public bool DumpsEnemy
        {
            get { return (BitConverter.ToUInt32(key, 0x800) | BitConverter.ToUInt32(key, 0x804)) != 0; }
        }
    }
}

File: KeySAVCoreJS/battle-video-reader.ts
import Pkx from "./pkx";
import * as util from "./util";

export default class BattleVideoReader {
    private static offset = 0x4E18;
    private static keyoff = 0x100;

    constructor(private video: Uint8Array, private key: Uint8Array) {}

    get dumpsEnemy() {
        return (this.key[0x800] | this.key[0x801] | this.key[0x802] | this.key[0x803]
              | this.key[0x804] | this.key[0x805] | this.key[0x806] | this.key[0x807]) != 0;
    }

    getPkx(slot: number, isOpponent: boolean): Pkx {
        var ekx;
        var pkx;
        var opponent: number = isOpponent ? 1 : 0;
        ekx = util.xor(this.video, BattleVideoReader.offset + 260 * slot + opponent * 1568,
            this.key, BattleVideoReader.keyoff + 260 * slot + opponent * 1792, 260);
        pkx = Pkx.decrypt(ekx);
        if (pkx.every(e => e == 0)) {
            return undefined;
        }
        return new Pkx(Pkx.verifyChk(pkx) ? pkx : ekx, -1, slot, false);
        }
}